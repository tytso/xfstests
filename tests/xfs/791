#! /bin/bash
# SPDX-License-Identifier: GPL-2.0-or-later
# Copyright (c) 2022 Oracle.  All Rights Reserved.
#
# FS QA Test No. 791
#
# Test scatter-gather atomic file writes.  We create a temporary file, write
# sparsely to it, then use XFS_EXCH_RANGE_FILE1_WRITTEN flag to swap
# atomicallly only the ranges that we wrote.  Inject an error so that we can
# test that log recovery finishes the swap.

. ./common/preamble
_begin_fstest auto quick fiexchange swapext

# Override the default cleanup function.
_cleanup()
{
	cd /
	rm -r -f $tmp.* $dir
}

# Import common functions.
. ./common/filter
. ./common/inject

# real QA test starts here
_supported_fs xfs
_require_xfs_io_command exchangerange
_require_xfs_scratch_atomicswap
_require_xfs_io_error_injection "bmap_finish_one"

_scratch_mkfs >> $seqres.full
_scratch_mount
_require_congruent_file_oplen $SCRATCH_MNT 65536

# Create original file
_pwrite_byte 0x58 0 1m $SCRATCH_MNT/a >> $seqres.full

# Create the donor file
$XFS_IO_PROG -f -c 'truncate 1m' $SCRATCH_MNT/b
_pwrite_byte 0x59 64k 64k $SCRATCH_MNT/b >> $seqres.full
_pwrite_byte 0x57 768k 64k $SCRATCH_MNT/b >> $seqres.full
sync

md5sum $SCRATCH_MNT/a | _filter_scratch
md5sum $SCRATCH_MNT/b | _filter_scratch

# Test swapext.  -h means skip holes in /b, and -e means operate to EOF
echo swap | tee -a $seqres.full
$XFS_IO_PROG -x -c 'inject bmap_finish_one' \
	-c "exchangerange -f -u -h -e $SCRATCH_MNT/b" $SCRATCH_MNT/a
_scratch_cycle_mount

md5sum $SCRATCH_MNT/a | _filter_scratch
md5sum $SCRATCH_MNT/b | _filter_scratch

# success, all done
status=0
exit
