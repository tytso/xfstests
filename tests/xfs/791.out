QA output created by 791
310f146ce52077fcd3308dcbe7632bb2  SCRATCH_MNT/a
c9fb827e2e3e579dc2a733ddad486d1d  SCRATCH_MNT/b
swap
exchangerange: Input/output error
e9cbfe8489a68efaa5fcf40cf3106118  SCRATCH_MNT/a
faf8ed02a5b0638096a817abcc6c2127  SCRATCH_MNT/b
